net/netcheck/netcheck.go

```diff
@@ -1441,7 +1461,23 @@ func (c *Client) addReportHistoryAndSetPreferredDERP(r *Report, dm tailcfg.DERPM
 	// still accessible and one of the conditions below is true.
 	keepOld := false
 	changingPreferred := prevDERP != 0 && r.PreferredDERP != prevDERP
-	oldRegionIsAccessible := oldRegionCurLatency != 0
+
+	// See if we've heard from our previous preferred DERP (other than via
+	// the STUN probe) since we started the netcheck, or in the past 2s, as
+	// another signal for "this region is still working".
+	heardViaOther := false
+	if changingPreferred && rs.opts != nil && rs.opts.GetLastDERPActivity != nil {
```

Comment:
consider adding this as a function on `opts` so that this can be abstracted away